pdns/syncres.hh

```diff
@@ -156,110 +152,84 @@ private:
 class DecayingEwma
 {
 public:
-  DecayingEwma() :  d_val(0.0)
-  {
-    d_needinit=true;
-    d_last.tv_sec = d_last.tv_usec = 0;
-    d_lastget=d_last;
-  }
-
+  DecayingEwma() {}
   DecayingEwma(const DecayingEwma& orig) = delete;
   DecayingEwma & operator=(const DecayingEwma& orig) = delete;
   
-  void submit(int val, const struct timeval* tv)
+  void submit(int val, const struct timeval& now)
   {
-    struct timeval now=*tv;
-
-    if(d_needinit) {
-      d_last=now;
-      d_lastget=now;
-      d_needinit=false;
+    if (d_last.tv_sec == 0 && d_last.tv_usec == 0) {
+      d_last = now;
       d_val = val;
     }
     else {
-      float diff= makeFloat(d_last - now);
-
-      d_last=now;
-      float factor=expf(diff)/2.0f; // might be '0.5', or 0.0001
-      d_val=(1-factor)*val + factor*d_val;
+      float diff = makeFloat(d_last - now);
+      d_last = now;
+      float factor = expf(diff)/2.0f; // might be '0.5', or 0.0001
+      d_val = (1-factor)*val + factor*d_val;
     }
   }
 
-  float get(const struct timeval* tv)
+  float get(float factor)
   {
```

Comment:
I understand that not computing the factor for every collection in a `DecayingEwmaCollection` is a nice optimization, but it feels like we are crippling the `DecayingEwma` class here, since it can no longer work outside of a `DecayingEwmaCollection`.

Reply:
True, but in this case I think it is worth it. 

It's not only avoiding the computation which produces the same value for every item in the collection, but also not having to store a `struct timeval` per item.

`DecayEwma` is only used for one purpose atm. If we encounter a case where it does make sense to store a timestamp per item it's easy to extend the class, adding a `struct timeval and a `get()` taking a now value and computing the factor.

But maybe the biggest benefit is that we have the `d_lastget` member available in the `DecayingEwmaCollection` class, making it more easy to define an index on that field in the future.

Reply:
Understood. It seems to me that `getFactor()` factor might then be better suited as a static method of `DecayingEwma` taking a `struct timeval` indicating the latest retrieval time and another one indicating the current time, but that's not important here.